src/envoy/http/authn/http_filter.cc

```diff
@@ -30,59 +57,211 @@ void AuthenticationFilter::onDestroy() {
-        } else {
-          ENVOY_LOG(error,
-                    "AuthenticationFilter: GetSourceIpPort() returns false!");
-        }
+        ENVOY_LOG(warn,
+                  "Source authentiation payload doesn't contain x509 nor jwt "
+                  "payload.");
       }
+    }
+
+    const auto& rule =
+        GetCredentialRuleOrDefault(context_.peer_user(), config_);
+    if (rule.origins_size() == 0) {
+      switch (rule.binding()) {
+        case iaapi::CredentialRule::USE_ORIGIN:
+          // Validation should reject policy that have rule to USE_ORIGIN but
+          // does not provide any origin method so this code should
+          // never reach. However, it's ok to treat it as authentication
+          // fails.
+          ENVOY_LOG(
+              warn,
+              "Principal is binded to origin, but not methods specified in "
+              "rule {}",
+              rule.DebugString());
+          onAuthenticateOriginDone(headers, &rule, 0, nullptr,
+                                   IstioAuthN::Status::FAILED);
+          break;
+        case iaapi::CredentialRule::USE_PEER:
+          // On the other hand, it's ok to have no (origin) methods if
+          // rule USE_SOURCE
+          onAuthenticateOriginDone(headers, &rule, 0, nullptr,
+                                   IstioAuthN::Status::SUCCESS);
+          break;
+        default:
+          // Should never come here.
+          ENVOY_LOG(error, "Invalid binding value for rule {}",
+                    rule.DebugString());
+          break;
+      }
+      return;
+    }
+    authenticateOrigin(
+        *headers, rule.origins(0),
+        std::bind(&AuthenticationFilter::onAuthenticateOriginDone, this,
+                  headers, &rule, 0, std::placeholders::_1,
+                  std::placeholders::_2));
+  }
+}
+
+void AuthenticationFilter::authenticateOrigin(
+    HeaderMap& headers, const iaapi::OriginAuthenticationMethod& method,
+    const AuthenticateDoneCallback& done_callback) {
+  ValidateJwt(headers, method.jwt(), done_callback);
+}
+
+void AuthenticationFilter::onAuthenticateOriginDone(
+    HeaderMap* headers, const iaapi::CredentialRule* rule, int method_index,
+    std::unique_ptr<IstioAuthN::AuthenticatePayload> payload,
+    const IstioAuthN::Status& status) {
+  if (status == IstioAuthN::Status::FAILED) {
+    // Try next one.
+    method_index++;
+    if (method_index < rule->origins_size()) {
+      authenticateOrigin(
+          *headers, rule->origins(method_index),
+          std::bind(&AuthenticationFilter::onAuthenticateOriginDone, this,
+                    headers, rule, method_index, std::placeholders::_1,
+                    std::placeholders::_2));
     } else {
-      ENVOY_LOG(
-          debug,
-          "AuthenticationFilter: {} this connection does not require mTLS",
-          __func__);
+      rejectRequest("Origin authentication failed.");
+    }
+  }
+  if (status == IstioAuthN::Status::SUCCESS) {
```

Comment:
For better readability and avoid a branch be entered in error, the code in this function should be organized as:
  if (status == IstioAuthN::Status::FAILED) {
....} else { //status == IstioAuthN::Status::SUCCESS) 
....
}

Reply:
done